### ISSUE TITLE
Field name mapping updateable from constructor arguments for product fulltext collection

### ISSUE THREAD

**Issue by sky-hub** [Smile-SA/elasticsuite#2103]
**Is your feature request related to a problem? Please describe.**
Sometimes you need to add custom filters with custom behavior which may require adding a new field mapping in the `\Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\Collection::$fieldNameMapping` array.
Currently there is no way in adding additional fields unless:
- rewriting collection 
- reimplement method \Smile\ElasticsuiteCatalog\Model\Layer\Filter\DecimalFilterTrait::apply at filter level to pass `$this->getFilterField()` instead of `$this->getAttributeModel()->getAttributeCode()`
- intercepting \Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\Collection::addFieldToFilter with a before plugin and change $field value

**Describe the solution you'd like**
Add a new argument to `\Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\Collection::__construct` named `$fieldNameMapping` of type array, which will be merged with class property `$fieldNameMapping`. This way new mappings can be configured from di.xml

**Describe alternatives you've considered**
A new class injected into `\Smile\ElasticsuiteCatalog\Model\ResourceModel\Product\Fulltext\Collection::__construct` with the sole purpose of handling field mappings?

**Additional context**
N/A

**Comment by romainruaud** [comment]
I agree on this, this has been on our todo from so long. But we ended up duplicating this `$fieldNameMapping` at many other locations (API, GraphQL, attributes rules, etc...).

So I'd be happy to review a PR that could fix this, with an unique object that could be also overriden/enriched through DI.

regards

**Comment by sky-hub** [comment]
@romainruaud , I will think of a solution in the following days and post it here so we can discuss it, if that is ok with you.
I guess it would be nice to offer some backwards compatibility also.

**Comment by romainruaud** [comment]
Yes @sky-hub feel free to propose a PR associated to this.

**Comment by sky-hub** [comment]
@romainruaud, this is next on my TODO list. I hope to have a PR soon.

**Comment by sky-hub** [comment]
@romainruaud, here's a draft PR: https://github.com/Smile-SA/elasticsuite/pull/2121

The PR only covers field mapping for search request builder, layer navigation filters, product fulltext collection and catalog rule attributes.

In regards of layer navigation filters I did not implement field mappings for 
\Smile\ElasticsuiteCatalog\Model\Layer\Filter\Attribute
\Smile\ElasticsuiteCatalog\Model\Layer\Filter\Boolean

as I do not see any benefit for having them here. If you think there could be cases when is needed I would like your opinion on how to handle 
\Smile\ElasticsuiteCatalog\Helper\AbstractAttribute::getFilterField 
and 
\Smile\ElasticsuiteCore\Helper\Mapping::getOptionTextFieldName

Is there any point in having a field mapping for a field like "option_text_attribute_code"?


There is one more area to cover which is the bucket aggregations. This area does not seem to need field mappings but instead some configuration more oriented to bucket naming and the field associated to the bucket.

**Comment by sky-hub** [comment]
@romainruaud , PR has been updated. 

PS: I am still waiting for your opinion on my last comment.
